### ISSUE TITLE
Pythia configuration

### ISSUE THREAD

**Issue by jstupak** [Snowmass21-software/MCProd#1]
Find where pythia is configured in Delphes and discuss this with Steve Mrenna

**Comment by trtomei** [comment]
The Pythia configuration that is standard in MadGraph sits in `$OUTPUTDIR/Cards/pythia8_card_default.dat` where `$OUTPUTDIR` is the whatever directory you told MadGraph to output to (using the `output` command just after `generate`). Stripping all the comments, it contains:

```
Main:numberOfEvents      = -1
HEPMCoutput:file         = auto
JetMatching:qCut         = -1.0
JetMatching:doShowerKt   = off
JetMatching:nJetMax      = -1
Merging:TMS              = -1.0
Merging:Process          = <set_by_user>
Merging:nJetMax  		 = -1
SysCalc:fullCutVariation = off
```

which means, it just takes all the Pythia 8 standard values.

**Comment by trtomei** [comment]
If we need a tune, the CMS CP5 tune (standard for Run 2) is described here:
http://dx.doi.org/10.1140/epjc/s10052-016-3988-x

and the (fully public) implementation is here:
https://github.com/cms-sw/cmssw/blob/master/Configuration/Generator/python/Pythia8CommonSettings_cfi.py
https://github.com/cms-sw/cmssw/blob/master/Configuration/Generator/python/MCTunes2017/PythiaCP5Settings_cfi.py

**Comment by jstupak** [comment]
Hello Thiago,

Thanks a lot for the info!

Regards,
John

**Comment by kkrizka** [comment]
I tried to get a working Pythia setup for #4. One caveat to @trtomei 's stripped down version of the default `pythia8_card.dat` is that it only works when using MG5's Py8 integration. MG5 program does the replacement of the "automatic" `-1` values to produce a usable card ([example](https://bazaar.launchpad.net/~madteam/mg5amcnlo/3.x/view/head:/madgraph/interface/madevent_interface.py#L3953)). You cannot just feed the default to `MG5aMC_PY8_interface` or one of the Pythia8 `main` programs.

Were you able to have MG5 run Py8 automatically inside gridpack?

I've been using the following, which I got by interrupting Py8 running inside MG5.  
```  
! Configure event input  
Beams:frameType=4  
Beams:LHEF=events.lhe.gz  
Beams:setProductionScalesFromLHEF=on  
LHEFInputs:nSubruns=1  
  
! Configure event output  
HEPMCoutput:file = events.hepmc  
  
! Matching configuration  
JetMatching:coneRadius = 1.0000000000e+00  
JetMatching:doShowerKt = off  
JetMatching:doVeto = off  
JetMatching:etaJetMax = 1.0000000000e+03  
JetMatching:merge = on  
JetMatching:nJetMax = 3  
JetMatching:nQmatch = 5  
JetMatching:qCut = 4.5000000000e+01  
JetMatching:scheme = 1  
JetMatching:setMad = off  
```

**Comment by jstupak** [comment]
I just brought this up in #10 .  I compiled gridpacks with shower=Pythia8, but when you run the gridpack it does not run pythia.  So this is a blocker for the actual production, and we need to figure this out

**Comment by jstupak** [comment]
this might be useful: https://answers.launchpad.net/mg5amcnlo/+question/231011

**Comment by kkrizka** [comment]
I don't think that works inside a gridpack. It seems to want a run output. I don't think the `run.sh` code registers the output as a "run". See the following:

```
Singularity> ls -l
total 11451
-rw-r--r-- 1 kkrizka snowmass21 11654833 Dec  7 16:25 events.lhe.gz   <---- made this by running `./run.sh 10000 0`
drwxr-xr-x 9 kkrizka snowmass21       14 Dec  7 16:39 madevent
-rwxr-xr-x 1 kkrizka snowmass21     2761 Dec  7 11:32 run.sh
Singularity> ./madevent/bin/madevent pythia8
launch in debug mode
************************************************************
*                                                          *
*                      W E L C O M E to                    *
*             M A D G R A P H 5 _ a M C @ N L O            *
*                      M A D E V E N T                     *
*                                                          *
*                 *                       *                *
*                   *        * *        *                  *
*                     * * * * 5 * * * *                    *
*                   *        * *        *                  *
*                 *                       *                *
*                                                          *
*         VERSION 3.3.0                 20xx-xx-xx         *
*                                                          *
*    The MadGraph5_aMC@NLO Development Team - Find us at   *
*    https://server06.fynu.ucl.ac.be/projects/madgraph     *
*                                                          *
*               Type 'help' for in-line help.              *
*                                                          *
************************************************************
INFO: load configuration from /home/kkrizka/.mg5/mg5_configuration.txt
INFO: load configuration from /work/MCProd/None/madevent/Cards/me5_configuration.txt
INFO: load configuration from /work/MCProd/None/madevent/Cards/me5_configuration.txt
No valid eps viewer found. Please set in ./input/mg5_configuration.txt
No valid web browser found. Please set in ./input/mg5_configuration.txt
pythia8
Command "pythia8" interrupted with error:
InvalidCmd : No run name currently define. Please add this information.
INFO:
quit
INFO:
```

**Comment by kkrizka** [comment]
Sorry. I think I mispoke. The above error is because I forgot to specify the run.

The `run.sh` does create a new run called `GridRun_{seed}`. It moves the events out of it though. But that can be undone.

Let me try something...

**Comment by kkrizka** [comment]
FYI: https://answers.launchpad.net/mg5amcnlo/+question/680362
The MadGraph authors say not to use the Py8 integration when running on a gridpack.

That being said, the following works if you do not `clean4grid`.
```
./bin/gridrun 1000 0 1
./bin/madevent pythia8 GridRun_0 -f
./bin/madevent delphes GridRun_0 -f
```

**Comment by kkrizka** [comment]
I've had varied success with the commands above over the last few days. I've switched to using the following lines for standalone Pythia8/Delphes. The paths are setup for my docker image. The pythia8_card is custom made following example above.

```sh
cd madevent
./bin/gridrun 10000 ${seed} 1
pushd Events/GridRun_${seed}
/opt/pythia/8.306/MG5aMC_PY8_interface ../../Cards/pythia8_card.dat
/opt/delphes/3.5.0/bin/DelphesHepMC2 ../../Cards/delphes_card_default.dat delphes.root events.hepmc
```

**Comment by trtomei** [comment]
@kkrizka Please link "your docker image", I want to try to use it as well!

**Comment by kkrizka** [comment]
Yup, that's the one. You should be able to get the latest one from `ghcr.io/kkrizka/mcprod:main`.

**Comment by jstupak** [comment]
@kkrizka regarding your example card:
> ```
> ! Configure event input  
> Beams:frameType=4  
> Beams:LHEF=events.lhe.gz  
> Beams:setProductionScalesFromLHEF=on  
> LHEFInputs:nSubruns=1  
>   
> ! Configure event output  
> HEPMCoutput:file = events.hepmc  
>   
> ! Matching configuration  
> JetMatching:coneRadius = 1.0000000000e+00  
> JetMatching:doShowerKt = off  
> JetMatching:doVeto = off  
> JetMatching:etaJetMax = 1.0000000000e+03  
> JetMatching:merge = on  
> JetMatching:nJetMax = 3  
> JetMatching:nQmatch = 5  
> JetMatching:qCut = 4.5000000000e+01  
> JetMatching:scheme = 1  
> JetMatching:setMad = off  
> ```

And comparing to CMS Cards:
https://github.com/cms-sw/cmssw/blob/master/Configuration/Generator/python/Pythia8CommonSettings_cfi.py
https://github.com/cms-sw/cmssw/blob/master/Configuration/Generator/python/MCTunes2017/PythiaCP5Settings_cfi.py

I see one conflict.  `Beams:setProductionScalesFromLHEF` is set to `on` in your example, and `off` in the CMS card.  I assume we want to stick with `on` here, but I am not sure.

And relative to what we have been using so far, you have `JetMatching:doShowerKt` set to `off`, but we have so far been setting this to `on`.  Again I'm not sure, but I think we want `on` based on this:
http://nuclear.ucdavis.edu/~christos/Pythia/pythia8226/share/Pythia8/phpdoc/JetMatching.php
> JetMatching:doShowerKt On Off   (default = off)
> This switch changes the merging prescription to the shower-kT matching scheme outlined in [Alw08]. This scheme differs from "classical" MLM jet matching with respect to when the matching veto is checked. The shower-kT scheme considers already immediately after the first shower emission if an event should be discarded. A detailed comparison of the "classical" MLM prescription (e.g. Madgraph-style matching with JetMatching:doShowerKt = off) and the shower-kT scheme are in preparation.

Ultimately we should run this card by Steve, but for now do you agree with my guesses?

**Comment by jstupak** [comment]
Hello @kkrizka .  I am trying something along the lines you suggested and I am getting an error:
`xqcut>0 but ickkw=0. Potentially not fully consistent setup. Be careful`

Did you see anything like this?

**Comment by jstupak** [comment]
I got around this by adding this to my pythia card: `JetMatching:setMad = 1`, which seems to instruct pythia to take the settings from LHE header (put there by madgraph). Pythia seems to accept all events I pass in.  But then when I run delphes, it is ~instantaneous and the output tree has zero events.  So still a wIP

**Comment by jstupak** [comment]
I think I have this working on my pythiaInGP branch.  If anyone wants to try, just run the `testGridPacks.sh` script.
